### PR TITLE
Сlient cookies synchronization with server (part of #1137, fix #1486, fix #1303, fix #1275, fix #1168)

### LINKED ISSUE #1168: Cookies synchronization before unload does not work in iOS
Browsers on iOS platform doesn't support `beforeunload` event.
We cann't use the `pagehide` event for cookie synchronization.
Request for a new page will be handled earlier than sync xhr request was sent from `pagehide` event handler.

### LINKED ISSUE #1275: Optimization of synchronization cookies with proxy server
Page sets cookies each second on https://www.salesforce.com/. Diference between cookies is expires time.
```
_kuid_=LcjSjrH1; domain=.krxd.net; path=/; expires=Sun, 18 Feb 2018 12:57:23 GMT
_kuid_=LcjSjrH1; domain=.krxd.net; path=/; expires=Sun, 18 Feb 2018 12:52:28 GMT
_kuid_=LcjSjrH1; domain=.krxd.net; path=/; expires=Sun, 18 Feb 2018 12:52:00 GMT
```

### LINKED ISSUE #1303: Periodic 'cookie-sync' service message hung the 'https://www.salesforce.com/' site
see [video](https://www.screencast.com/t/2pfs90vhEEj)

### LINKED ISSUE #1486: Cookie synchronization with server is too slow
The xhr request may outrun the cookie synchronization and send request to destination server without cookie. But it can be if client script sets a cookie right before it will send a request.

### PR DESCRIPTION
**Pull request by LavrovArtem**

### CODE REVIEW

File: test/client/fixtures/utils/cookie-test.js [hunk 1]
```diff
@@ -72,6 +72,7 @@ test('formatClientString', function () {
 });
 
 test('domainMatch', function () {
+    ok(cookieUtil.domainMatch('sub.example.com'));
```
Comment (miherlosev) [inline test/client/fixtures/utils/cookie-test.js]: You call this function only with two arguments. So test it with two arguments as well: `ok(cookieUtil.domainMatch('sub.example.com', void 0))`.

For `pathMatch` too.

File: src/client/sandbox/cookie/index.js [hunk 2]
```diff
@@ -118,32 +91,38 @@ export default class CookieSandbox extends SandboxBase {
...
-            this._updateClientCookieStr(parsedCookie.key, clientCookieStr);
+            CookieSandbox._updateClientCookieStr(parsedCookie.key, clientCookieStr);
         }
 
-        // NOTE: Meanwhile, synchronize cookies with the server cookie jar.
-        if (syncWithServer)
-            // eslint-disable-next-line no-restricted-properties
-            this.cookieSync.perform({ url: document.location.href, cookie });
+        if (setByClient) {
+            cookieUtils.setDefaultValues(parsedCookie, parsedDestLocation);
```
[the remark this answers is not in the document]
Comment (miherlosev) [reply src/client/sandbox/cookie/index.js]: Ok. Let it be `cookieUtils.setDefaultValues(parsedCookie, { hostname, pathname })`

**Comment by LavrovArtem** [comment]
@testcafe-build-bot retest

File: src/client/utils/destination-location.js [hunk 3]
```diff
@@ -79,7 +79,7 @@ export function getParsed () {
     let pathname   = nativeMethods.anchorPathnameGetter.call(resolver);
 
     // NOTE: IE ignores the first '/' symbol in the pathname.
-    if (pathname.indexOf('/') !== 0)
+    if (pathname.charAt(0) !== '/')
```
Comment (AndreyBelym) [inline src/client/utils/destination-location.js]: What's the reason for not using `pathname[0]`?

Comment (LavrovArtem) [reply src/client/utils/destination-location.js]: I don't know :)
Usually the `charAt` is used in the [testcafe](https://github.com/DevExpress/testcafe/search?q=charAt&unscoped_q=charAt) and [hammerhead](https://github.com/DevExpress/testcafe-hammerhead/search?q=charAt&unscoped_q=charAt).

Comment (kirovboris) [reply src/client/utils/destination-location.js]: old IE support?

**Pull request merged**